execution-plane/arcon/src/stream/operator/mod.rs

```diff
@@ -15,17 +18,56 @@ where
     IN: ArconType,
     OUT: ArconType,
 {
+    /// Called by parent node to finish the initialization of the operator
+    fn init(&mut self, _state_backend: &mut dyn StateBackend) {}
+
     /// Determines how the `Operator` processes Elements
     ///
-    /// The function takes an Element and a [ChannelStrategy] in order to pass on it
-    fn handle_element(&mut self, element: ArconElement<IN>, strategy: &mut ChannelStrategy<OUT>);
+    /// The function takes an Element and a [NodeContext] in order to pass on it
+    fn handle_element(&mut self, element: ArconElement<IN>, ctx: OperatorContext<OUT>);
+
     /// Determines how the `Operator` processes Watermarks
     ///
     /// The function either returns None or a Vec of ArconEvents (e.g., Window Computation)
-    fn handle_watermark(&mut self, watermark: Watermark) -> Option<Vec<ArconEvent<OUT>>>;
+    fn handle_watermark(
+        &mut self,
+        watermark: Watermark,
+        ctx: OperatorContext<OUT>,
+    ) -> Option<Vec<ArconEvent<OUT>>>;
+
     /// Determines how the `Operator` processes an Epoch marker
     ///
     /// The function either returns None signaling it did not attempt to snapshot any state.
     /// If the `Operator` snapshotted its state, the raw bytes are packed into an `ArconResult`
-    fn handle_epoch(&mut self, epoch: Epoch) -> Option<ArconResult<Vec<u8>>>;
+    fn handle_epoch(
+        &mut self,
+        epoch: Epoch,
+        ctx: OperatorContext<OUT>,
+    ) -> Option<ArconResult<Vec<u8>>>;
+}
+
+pub struct OperatorContext<'c, 's, OUT: ArconType> {
+    channel_strategy: &'c mut ChannelStrategy<OUT>,
+    pub state_backend: &'s mut dyn StateBackend,
```

Comment:
Wondering how we'll deal with stateless operators/nodes. They won't need any StateBackend. If this should be an option or something.

Reply:
Doesn't necessarily hurt to give them one anyway, even if they are ignoring it, does it?

Reply:
`Node` always has its own `StateBackend` though. It can pass it without any problems to the operator inside, even though the operator doesn't use it.

Reply:
Well, I guess not ^^